hydra-node/src/Hydra/Events/Rotation.hs

```diff
@@ -0,0 +1,104 @@
+module Hydra.Events.Rotation where
+
+import Hydra.Prelude
+
+import Conduit (MonadUnliftIO)
+import Control.Concurrent.Class.MonadSTM (newTVarIO, readTVarIO, writeTVar)
+import Hydra.Chain.ChainState (ChainStateType, IsChainState)
+import Hydra.Events (EventSink (..), EventSource (..), HasEventId, LogId, StateEvent (..), getEvents)
+import Hydra.HeadLogic (StateChanged (Checkpoint), aggregate)
+import Hydra.HeadLogic.State (HeadState (..), IdleState (..))
+
+newtype RotationConfig = RotateAfter Natural
+
+-- | An EventSource and EventSink combined
+type EventStore e m = (EventSource e m, EventSink e m)
+
+type Checkpointer e = [e] -> e
+
+-- | Creates an event store that rotates according to given config and 'Checkpointer'.
+mkRotatedEventStore ::
+  (HasEventId e, MonadSTM m, MonadUnliftIO m) =>
+  RotationConfig ->
+  Checkpointer e ->
+  LogId ->
+  EventStore e m ->
+  m (EventStore e m)
+mkRotatedEventStore config checkpointer logId eventStore = do
+  logIdV <- newTVarIO logId
+  -- Rules for any event store:
+  -- - sourceEvents will be called in the beginning of the application and whenever the api server wans to load history
+  --   -> might be called multiple times!!
+  -- - putEvent will be called on application start with all events returned by sourceEvents and during processing
+  currentEvents <- getEvents eventSource
+  let currentNumberOfEvents = toInteger $ length currentEvents
+  numberOfEventsV <- newTVarIO currentNumberOfEvents
+  -- XXX: check rotation on startup
+  when (currentNumberOfEvents >= toInteger rotateAfterX) $ do
+    rotateEventLog logIdV numberOfEventsV
```

Comment:
I don't think we need this as the application will call `sourceEvents` at least once (see rules above).

So indeed we should hook into the `sourceEvents` conduit to do our event counting (and maybe do rotation while sourcing!?)

Reply:
Please note [this](https://github.com/cardano-scaling/hydra/pull/1997/files#diff-9f9f50503db37faf1f8b51d54169742a9751a16583b6f5cbbe264ea00322fe85R46) decision prevents making the "rotation on startup" logic outside this function (e.g., in hydrate)

Reply:
As per our latest discussion, we decided to move forward with the current approach, as rotating while sourcing is more complex.

 Given this is a first attempt, we prefer to keep it as simple as possible.

That said, we may revisit and adopt that strategy in a future follow-up.